### DISCUSSION TITLE
APIs for TestFlight Feedback (Crashes and Screenshots)

### DISCUSSION THREAD

**Discussion by erikkerber** [fastlane/fastlane#17790] [Ideas]
Not seeing any API, or discussion, for retrieving TestFlight Feedback via Spaceship I thought I'd start a thread here.

We're leveraging TestFlight pretty heavily now, with thousands of Beta testers. The feedback feature is great, but it's slow as a dog and requires one to have the appropriate Connect role. We'd prefer to marshal this data to an internal tool, or to be frankly honest live-post it to Slack so its more accessible and easier to browse and reference.

<img width="155" alt="CleanShot 2020-12-13 at 12 07 33@2x" src="https://user-images.githubusercontent.com/75236/102019952-d7557980-3d3b-11eb-8b41-daecc75a43c6.png">

Is the lack of support for accessing this information difficult to do programmatically? Or has it just not happened yet? If the latter, I may be interested in diving in but don't necessarily want to do that if this is a known rabbit hole to avoid.

The URL for screenshots take an `appid` (easy), and a pre-release `versionid` (easy?).

```
https://appstoreconnect.apple.com/apps/<appid>/testflight/screenshots?preReleaseVersionId=b550dc76-875d-4325-8eba-85ea9a2f117d&sort=-timestamp
```

It looks like the screenshot feedback is accessible via JSON response as well, so no web scraping needed:

```
https://appstoreconnect.apple.com/iris/v1/betaFeedbacks?filter[build.app]=<appid>&filter[build.preReleaseVersion]=<buildid>&sort=-timestamp&exists[crash]=false&include=tester,build,screenshots,buildBundle&fields[betaTesters]=firstName,lastName&fields[builds]=version&limit=60
```

The response is large, and the images and comments aren't globbed together in one object, so some mapping and reducing will be necessary.

Pending no red lights we might dig into it.

**Comment by lukas-ruzicka** [comment]
@erikkerber Do you know how to authenticate the `https://appstoreconnect.apple.com/iris/v1/betaFeedbacks` endpoint? It's not possible with the JWT token as for other App Store Connect API calls

**Comment by abardallis** [comment]
@lukas-ruzicka we've had some success using `fastlane spaceauth` to generate a `FASTLANE_SESSION` env variable. 
See Method 2 on [this documentation page](https://docs.fastlane.tools/getting-started/ios/authentication/) for how to generate that.

With a valid `FASTLANE_SESSION` env variable in place, we've cooked up some Ruby scripts that call `Spaceship::ConnectAPI` methods directly. For instance, in the method we use to fetch feedback we have something closely resembling the following:

```ruby
response = Spaceship::ConnectAPI.get_beta_feedback(filter: {
                                                  'build.app' => app_id,
                                                  'build' => build_id
                                                },
                                                includes: 'tester,build,screenshots',
                                                sort: '-timestamp').all_pages
feedbacks = response.flat_map do |r|
  r.body
end
```

### Tips:
- These methods aren't documented too well for use in this way, but dig around in the `spaceship`/`connect_api` source code enough and you'll start seeing all sorts of different calls you can make and what arguments they accept.
- Use `irb` a ton to experiment.

From the fastlane docs (talking about `FASTLANE_SESSION`):
> The session's validity can greatly vary (anything between 1 day and 1 month, depending on factors such as geolocation of the session usage). This means you'll have to generate a new session at least once a month. Usually you'd only know about it when your build starts failing.

- So keep logs of successful/failed fetches. You'll notice your data will stop updating. You'll go check your logs and see failures. 99% of this time, this means that your `FASTLANE_SESSION` needs refreshing.
- After you get this working, you'll start asking yourself how you can automate the `FASTLANE_SESSION` generation process given that it still involves 2FA codes. We tried to get as close to this as we could (short of using twilio... which we even considered). We even used some `expect` scripting and the Slack gem to handle passing requests and 2FA codes into the non-interactive shell running `fastlane spaceauth`. All I can say is, create your own Rube Goldberg machine at your own risk. My life got better when we simplified.

**Comment by DoodySq** [comment]
Thanks for the information, I will try to figure it out for more. Keep sharing such informative post keep suggesting such post.


[Quest Diagnostics Appointment
](https://www.questdiagnosticsbill.me/)

**Comment by lukas-ruzicka** [comment]
Hey guys, many thanks for your inputs here. We’d done some work on this topic and created a tool that fetches the TestFlight feedback and creates GitHub Issues out of it. I’ve made it open-source, so you can use it straight away if you want - it’s built as a GitHub Action, so all you need to do is to create a GitHub repository and add workflow files there (follow the instruction in README). Hope this helps!

If you want other integrations or have any ideas for improvements, please feel free to contribute. 

https://github.com/lukas-ruzicka/testflight-feedback

**Comment by mhyassin** [comment]
Hey guys, the recent action didn't work for me, I've created a new typescript one 

https://github.com/mhyassin/testflight-feedback-import
